### ISSUE TITLE
Suggestion to improve XML shredding performance 

### ISSUE THREAD

**Issue by micke314** [erikdarlingdata/DarlingData#31]
From what I can tell, #human_events_xml ends up with 1 row with entire XML. If possible it is often a good thing to pre-shred that to smaler parts. Here it looks like you could change #human_events_xml to have 1 row for each event instead and do this change without changing any other code and all will hopefully be fast and sweet.
For me with a session capturing about 50 queries the part that shreds the XML went from 4 seconds to 0.8 seconds.
Would be really interesting to see if you see the same. Perhaps you have other problem areas to test?

Sugested code change filling #human_events_xml

```
--Dump whatever we got into a temp table
declare @X xml;

select @X = convert(xml, t.target_data)
from sys.dm_xe_session_targets as T
  join sys.dm_xe_sessions as s
    on S.address = T.event_session_address
where S.name = @session_name 
  and T.target_name = N'ring_buffer' 
option (recompile);

select E.X.query('.') as human_events_xml
into #human_events_xml
from @X.nodes('/RingBufferTarget/event') as E(X)
option (recompile);

```

**Comment by erikdarlingdata** [comment]
Just to make sure I fully understand: what's the advantage of assigning the XML to a variable up front, and then shredding? Does something weird happen in the DMV query that could lead to extra XML-shredding before the `target_name` filter is applied?

**Comment by micke314** [comment]
You mean instead of shredding on the value converted to xml from nvarchar(max) in the DMV. That is blogged about here. 

https://www.sql.kiwi/2012/09/compute-scalars-expressions-and-execution-plan-performance.html

The cast to XML is deferred to where ever the query needs the XML, for each row, for each value and when the cast is done the XML is validated over and over again.

So by casting to a variable the cast happens only once.

**Comment by erikdarlingdata** [comment]
Oh wow, that's badass! Thanks. I wonder if I could use that in sp_BlitzLock too.

**Comment by micke314** [comment]
It does not of course have to be a variable to fix this, it could as well have been to a temp table with an XML column.

You will have the exact same issue if you shred queryplans from the query store.
Dump the querystore to a temp table with an XML column and shred  the temp table instead.
Will probably be much faster.

**Comment by erikdarlingdata** [comment]
@micke314 works awesome! Thank you!
